### ISSUE TITLE
Lifecycle hook before removing elements

### ISSUE THREAD

**Issue by Exocomp** [dhilt/ngx-ui-scroll#222]
Would you be able to expose a lifecycle hook which reports on the elements which are about to be removed as the user scrolls. Reason being currently Angular does not have a "ngBeforeDestroy" which means that by the time a components "ngOnDestroy" is triggered the element is already removed from the DOM.

By creating a "onBeforeDestroy" and reporting on the items about to be removed it gives a chance for the developer to do any final work prior to the items being removed. For example, it could be a promise which waits until it resolves then removes.

**Comment by dhilt** [comment]
@Exocomp This sounds very interesting and a bit weird. Could you provide an example of how you would use this API as if it was already implemented? Including all the necessary infrastructure (template, datasource, component BL). I want to explore all sides of the use case before thinking about implementation. The changes you want might be quite deep, and we need to be precisely consistent. Also, there is always a chance to find a specific workaround.

**Comment by Exocomp** [comment]
A work around is fine of course with existing functionality if that is possible.

Say for example I have the following:

```
<div *uiScroll="let item of datasource">
  <app-my-component></app-my-component>
</div>
```

app-my-component has view refs which I would like to detach before they are destroyed. See [here](https://github.com/angular/angular/issues/17404) for other developers looking for an Angular solution of having a ngBeforeDestory that would be the ideal way I think once it is implemented in Angular. However, in the mean time a work around to that was what I was thinking. Essentially if there is a method that is envoked before the items are removed (essentially resulting in the the items removed from the DOM) would be a potential solution.

So for example.

```
this.datasource = new Datasource({
  get: async (index, count, success) => {   
    ...
  },
  settings: {
    ...
    onBeforeDestroy: myHandler()
  }
});
```

What I was thinking is that myHandler() gets invoked with an array of items that are about to be removed. That gives a chance to detach the view refs from the component before the component is removed from the DOM.

**Comment by dhilt** [comment]
@Exocomp So you want the handler to be synchronous... this should be somewhat easier than I thought before. In the ngx-ui-scroll lib we have so called UIScroll Workflow which is a list of processes managed by a state machine. The very high-level image of this Workflow [is here](https://github.com/dhilt/ngx-ui-scroll/wiki/The-UIScroll-Workflow). We are interested in the part called Inner Loop, which is a bunch of main processes: 

PreFetch -> Fetch (async) -> PostFetch -> Render (async) -> PreClip -> Clip -> Adjust (async)

Some of them can be skipped during a single cycle, "async" means that the process **can be** asynchronous. Removing of items takes place inside [the Clip process](https://github.com/dhilt/ngx-ui-scroll/blob/v1.9.0-rc.1/src/component/processes/clip.ts), which is synchronous. So the thing we need to do is just:

 - arrange new setting (a method)
 - invoke this method when Clip (synchronously before changing the DOM)

I see, you forked the repo, what if I suggest you to draft an implementation? I will provide the roadmap, review the results and merge it when it is done. What do you think?

**Comment by dhilt** [comment]
@Exocomp The roadmap seems pretty straightforward. 

**1.** Add a new setting to the [interface](https://github.com/dhilt/ngx-ui-scroll/blob/v1.9.0-rc.1/src/component/interfaces/settings.ts) and to the [class](https://github.com/dhilt/ngx-ui-scroll/blob/v1.9.0-rc.1/src/component/classes/settings.ts#L18-L19). Let it be `onBeforeClip?: Function`.

**2.** Set new Input: [enum](https://github.com/dhilt/ngx-ui-scroll/blob/v1.9.0-rc.1/src/component/inputs/settings.ts#L19) and [type](https://github.com/dhilt/ngx-ui-scroll/blob/v1.9.0-rc.1/src/component/inputs/settings.ts#L91).

```ts
  [Settings.onBeforeClip]: {
    validators: [FUNC],
    defaultValue: void 0
  },
```

**3.** Go to Clip and call the method (if present) in appropriate time. 

The point is that the `Clip` process does not perform DOM-removing. It [hides the elements](https://github.com/dhilt/ngx-ui-scroll/blob/v1.9.0-rc.1/src/component/processes/clip.ts#L32),  and [removes the items](https://github.com/dhilt/ngx-ui-scroll/blob/v1.9.0-rc.1/src/component/processes/clip.ts#L39-L43) from the data array bound with the template. This way the Clip process is synchronous, and real removing from DOM happens later, but I believe it doesn't have an impact on what we are going to do.

So my suggestion is to invoke the handler [between DOM-hiding and data-removing](https://github.com/dhilt/ngx-ui-scroll/blob/v1.9.0-rc.1/src/component/processes/clip.ts#L38):

```ts
if (scroller.settings.onBeforeClip) {
  scroller.settings.onBeforeClip(itemsToRemove);
}
```

This is how I see it right now. Of course, it might be not the end, but I don't think we'll be stuck here. Can you try it in your environment and give a feedback? I would be happy to continue the discussion in your Pull Request.

**Comment by Exocomp** [comment]
Thanks for the details, it is working well with my initial testing. I'll make the change to my project and do some more testing and see how well it plays out and will submit the PR.

**Comment by Exocomp** [comment]
Here is the PR [223](https://github.com/dhilt/ngx-ui-scroll/pull/223).

**Comment by dhilt** [comment]
@Exocomp Fine job! I pushed [few changes over your code](https://github.com/dhilt/ngx-ui-scroll/commit/5ef4030b8ed1a214b9bcad7f548c0cfbe660b2e3) and opened another [PR related to spaces problem](https://github.com/robisim74/angular-library-starter/pull/70). Also, I published ngx-ui-scroll v1.9.0-rc.2 for not to keep you waiting. This feature will be included into 1.9.0 with a demo and appropriate documentation. But you already may use 1.9.0-rc.2.

**Comment by Exocomp** [comment]
Nice, thank you!

**Comment by dhilt** [comment]
@Exocomp I released ngx-ui-scroll v1.9.0, the API has changed a bit, the argument is the list of _ItemAdapter_ objects now:

```js
  onBeforeClip: (items: {
    $index: number,
    data: any,
    element?: HTMLElement
  }[]) => void
```

Doc: https://dhilt.github.io/ngx-ui-scroll/#/experimental#onBeforeClip-setting
